### PR TITLE
feat: support tls expression route.

### PR DESCRIPTION
**Pull request by rodman10**
**What this PR does / why we need it**:

**Which issue this PR fixes**:

Fixes [#4540](https://github.com/Kong/kubernetes-ingress-controller/issues/4540)

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/dataplane/parser/translate_tlsroute.go [hunk 1]
```diff
@@ -29,11 +29,6 @@ func (p *Parser) ingressRulesFromTLSRoutes() ingressRules {
 
 	var errs []error
 	for _, tlsroute := range tlsRouteList {
-		if p.featureFlags.ExpressionRoutes {
-			p.registerResourceFailureNotSupportedForExpressionRoutes(tlsroute)
-			continue
-		}
-
 		if err := p.ingressRulesFromTLSRoute(&result, tlsroute); err != nil {
```
Comment (randmonkey) [inline internal/dataplane/parser/translate_tlsroute.go:32]: Need to check kong version from parser here. If the KongVersion is before `ExpressionRouterL4Cutoff` (3.4.0), Kong does not support L4 proxy with expression router, so we still should register a translation failure here.

**Comment by rodman10** [comment]
I have struggled with the tls_passthrough `intergration` test all day, I found [traditional flavor](https://github.com/Kong/kong/blob/master/kong/router/traditional.lua#L923-L1146) hadn't constrainted the `net.protocol` for `stream` subsystem strictly, so the `tls_passthrough` which will be converted into `tcp` will work properly, but in expression flavor, the conversion of `tls_passthrough` will cause expression match failed for gateway appending `protocols` field to the expression (for `tls_passthrough` is `tcp`) , but the `scheme` (aka `protocols`) will be set to `tls` in gateway `preread` phase, which will cause route not found, whether I have missed or misunderstanding anything? @randmonkey

File: internal/dataplane/parser/translate_tlsroute.go (hunk 1 above)
Comment (rodman10) [reply internal/dataplane/parser/translate_tlsroute.go:32]: Ok, I will fix it.

**Comment by randmonkey** [comment]
@rodman10 confirmed with gateway team that there are some issue in the Kong router part to set the matching protocols, so `tls_passthrough` is not supported now. The Kong's PR https://github.com/Kong/kong/pull/11538 is expected to fix this.

**Comment by rodman10** [comment]
> @rodman10 confirmed with gateway team that there are some issue in the Kong router part to set the matching protocols, so `tls_passthrough` is not supported now. The Kong's PR [Kong/kong#11538](https://github.com/Kong/kong/pull/11538) is expected to fix this.

Whether `tls_pasthrough` should be contained in this PR, or according to this [thread](https://github.com/Kong/kubernetes-ingress-controller/issues/4540), I can just ignore the test failure? @randmonkey

**Comment by randmonkey** [comment]
> > @rodman10 confirmed with gateway team that there are some issue in the Kong router part to set the matching protocols, so `tls_passthrough` is not supported now. The Kong's PR [Kong/kong#11538](https://github.com/Kong/kong/pull/11538) is expected to fix this.
> 
> Whether `tls_pasthrough` should be contained in this PR, or according to this [thread](https://github.com/Kong/kubernetes-ingress-controller/issues/4540), I can just ignore the test failure? @randmonkey
I think we should hold this until Kong gateway has released a version to support `tls_passthrough`. Before Kong supports this, we cannot achieve the support of `TLSRoute`. While in your implementation, I think it is OK to ignore the test failures for now.

**Comment by chronolaw** [comment]
> > @rodman10 confirmed with gateway team that there are some issue in the Kong router part to set the matching protocols, so `tls_passthrough` is not supported now. The Kong's PR [Kong/kong#11538](https://github.com/Kong/kong/pull/11538) is expected to fix this.
> 
> Whether `tls_pasthrough` should be contained in this PR, or according to this [thread](https://github.com/Kong/kubernetes-ingress-controller/issues/4540), I can just ignore the test failure? @randmonkey

Sorry, There is a typo in my PR(https://github.com/Kong/kong/pull/11538), it should be `tls_pasthough`, not `tls_pasthrough`.

**Comment by randmonkey** [comment]
@rodman10 Now Kong 3.4.1 is released with the fix of supporting TLS passthrough. You can continue on this PR if you still want to do.

**Comment by rodman10** [comment]
Ok, I will work with this pr this weekend.

**Comment by randmonkey** [comment]
@rodman10 Is it ready for review now? I tried to run integration tests for TLSRoute and the tests passed.

**Comment by rodman10** [comment]
> @rodman10 Is it ready for review now? I tried to run integration tests for TLSRoute and the tests passed.

Ok, the pr is ready for review, because I found kongintergration test was failed. I was trying to figure it out.

File: CHANGELOG.md [hunk 2]
```diff
@@ -142,6 +142,9 @@ Adding a new version? You'll need three changes:
 
 ### Added
 
+- Added support for expression-based Kong routes for `TLSRoute`. This requires the `ExpressionRoutes`
+  feature gate and a Kong 3.4.2 install with `KONG_ROUTER_FLAVOR=expressions` set in the environment.
```
Comment (randmonkey) [inline CHANGELOG.md:146]: Maybe we do not need to mention version of Kong here since KIC 3.0 requires Kong >= `3.4.1` ? @programmer04

Comment (rodman10) [reply CHANGELOG.md:146]: @randmonkey So I remove the version requirement here?

Comment (randmonkey) [reply CHANGELOG.md:146]: Yes, I think 
```suggestion
  feature gate and a Kong installed with `KONG_ROUTER_FLAVOR=expressions` set in the environment.
```
like this is OK. And please limit a line to contain at most 80 characters.

**Review by randmonkey — APPROVED** [review]

**Pull request merged**